allennlp/modules/span_pruner.py

```diff
@@ -0,0 +1,94 @@
+from typing import Callable, Tuple
+
+from overrides import overrides
+import torch
+
+from allennlp.nn import util
+from allennlp.common.checks import ConfigurationError
+
+class SpanPruner(torch.nn.Module):
+    """
+    This module scores and prunes span based representations using a parameterised scoring
+    function and a threshold.
+
+    Parameters
+    ----------
+    scorer : ``Callable[[torch.FloatTensor], torch.FloatTensor]``, required.
+        A function which, given a tensor of shape (batch_size, num_spans, embedding_size),
+        produces a tensor of shape (batch_size, num_spans, 1), representing a scalar score
+        per span in the tensor.
+    """
+    def __init__(self, scorer: Callable[[torch.FloatTensor], torch.FloatTensor]) -> None:
+        super(SpanPruner, self).__init__()
+        self._scorer = scorer
+
+    @overrides
+    def forward(self, # pylint: disable=arguments-differ
+                span_embeddings: torch.FloatTensor,
+                span_mask: torch.LongTensor,
+                num_spans_to_keep: int) -> Tuple[torch.FloatTensor, torch.LongTensor,
+                                                 torch.LongTensor, torch.FloatTensor]:
+        """
+        The indices of the top-k scoring spans according to span_scores. We return the
+        indices in their original order, not ordered by score, so that we can rely on
+        the ordering to consider the previous k spans as antecedents for each span later.
+        Parameters
+        ----------
+        span_embeddings : ``torch.FloatTensor``, required.
+            A tensor of shape (batch_size, num_spans, embedding_size), representing
+            the set of embedded span representations.
+        num_spans_to_keep : ``int``, required.
+            The number of spans to keep when pruning.
+
+        Returns
+        -------
+        top_span_embeddings : ``torch.FloatTensor``, required.
+            The span representations of the top-k scoring spans.
+            Has shape (batch_size, num_spans_to_keep, embedding_size).
+        top_span_mask : ``torch.LongTensor``, required.
+            The coresponding mask for ``top_span_embeddings``. Has shape
+            (batch_size, num_spans_to_keep).
+        top_span_indices : ``torch.IntTensor``, required.
+            The indices of the top-k scoring spans. Has shape (batch_size, num_spans_to_keep).
+        top_span_scores : ``torch.FloatTensor``, required.
+            The values of the top-k scoring spans. Has shape (batch_size, num_spans_to_keep, 1).
```

Comment:
Ugh yeah I know, the current coref model is very annoying for this. I started to change it, but it's big enough to be in it's own PR afterwards because it meant making a lot of the really confusing functions like `batch_index_select` support tensors which don't have the same shape.